### DISCUSSION TITLE
Reuse value generated by arbitrary multiple times

### DISCUSSION THREAD

**Discussion by mrmarbury** [dubzzz/fast-check#3074] [Q&A] [answered]
I feel that this is such a basic thing that it might have been documented or answered already but I can't for the love of god find anything.

What I do ... or try to do: I generate records containing multiple fields. But two fields need the exact same random string because they need to be matched through the code. 

Take this extremely simplified and probably stupid example:
```
export const buildArbitraryObject = (): fc.Arbitrary<MyType> => {
  return fc.record<MyType>({
    key: RANDOM_STRING,
    value: fc.record<SubType>({
      something: RANDOM_STRING
    })
  }),
}
```
I need to generate this RANDOM_STRING once but use the one generated value at two places. I tried to do 

`const randomString = fc.string({minLength: 3})`

und put the randomString const where I need that value but `fc.string()` is still run multiple times. 
How do I generate a random string once per generated example and use it at multiple places?

Thanks for help.

**Comment by dubzzz** [comment, marked as answer]
Option 1: not my recommendation as `chain` tend to shrink badly

```js
export const buildArbitraryObject = (): fc.Arbitrary<MyType> => {
  return fc.string().chain(s => fc.record<MyType>({
    key: s,
    value: fc.record<SubType>({
      something: s
    })
  })),
}
```

Option 2: with `map`

```js
export const buildArbitraryObject = (): fc.Arbitrary<MyType> => {
  return fc.record({
    key: s,
    value: fc.record({ // value without the something part
    })
  })).map(o => ({...o, value: {...o.value, something: o.key}})),
}
```

Option 3: I have to think a way to simplify the construction of such arbitrary to have something as simple as `chain` version (or even simpler) but with the shrink capacities of `map`

_Disclaimer: answered from my phone_

**Reply by mrmarbury** [reply]
Thank you very much, that helps a lot :) Regarding Option 3 and what I was looking for when I first encountered the issue: I was looking for a function like `value()` or similar that I could call like so `fc.string().value()`or similar to retrieve the value and store it in a variable. I don't know if that's possible though? I will start with the `chain()` version for now and move to `map()` when I run into problems.
